### PR TITLE
Use XmlWriter.Create instead of XmlTextWriter

### LINKED ISSUE #1411: `System.Xml.Serialization.XmlSerializer` serializes 0x00 but crashes on deserialization
sample
```csharp
public class StringContainer
{
    public string String { get; set; }
}

[Fact]
public void Test()
{
    var serializer = new XmlSerializer(typeof(StringContainer));

    var expected = new StringContainer
    {
        String = "Sample\0Text"
    };

    var stream = new MemoryStream();
    serializer.Serialize(stream, expected);

    stream.Position = 0;
    var actual = (StringContainer)serializer.Deserialize(stream); // <-- exception
}
```
it would be nice, if the exception would occur at serialization, because its a pita when it happens at deserialization with huge streams (ex `There is an error in XML document (54563, 98).`)

### PR DESCRIPTION
**Pull request by StephenMolloy**
Use XmlReader/Writer.Create instead of XmlTextReader/Writer.

XmlTextWriter is not strict about some of the things it writes. Like
the null character, which is technically illegal in XML 1.0 and 1.1.
(The way we used the XmlTextReader counterpart, we would end
up successfully serializing a null character, and then hit an exception
when trying to deserialize what we just serialized. This inability to
round trip breaks a core tenet of serialization.)

Since the long ago days of .Net 2.0, Microsoft has been
recommending people to use XmlReader/Writer.Create()
instead of directly instantiating XmlTextReader/Writer. This PR
will bring us in line with those recommendations from long ago.

Fixes #1411

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 1]
```diff
@@ -317,9 +317,10 @@ public void Serialize(TextWriter textWriter, object? o)
         [RequiresUnreferencedCode(TrimSerializationWarning)]
         public void Serialize(TextWriter textWriter, object? o, XmlSerializerNamespaces? namespaces)
         {
-            XmlTextWriter xmlWriter = new XmlTextWriter(textWriter);
-            xmlWriter.Formatting = Formatting.Indented;
-            xmlWriter.Indentation = 2;
+            //XmlTextWriter xmlWriter = new XmlTextWriter(textWriter);
+            //xmlWriter.Formatting = Formatting.Indented;
+            //xmlWriter.Indentation = 2;
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:322]:
```suggestion
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 2]
```diff
@@ -332,9 +333,10 @@ public void Serialize(Stream stream, object? o)
         [RequiresUnreferencedCode(TrimSerializationWarning)]
         public void Serialize(Stream stream, object? o, XmlSerializerNamespaces? namespaces)
         {
-            XmlTextWriter xmlWriter = new XmlTextWriter(stream, null);
-            xmlWriter.Formatting = Formatting.Indented;
-            xmlWriter.Indentation = 2;
+            //XmlTextWriter xmlWriter = new XmlTextWriter(stream, null);
+            //xmlWriter.Formatting = Formatting.Indented;
+            //xmlWriter.Indentation = 2;
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:338]:
```suggestion
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 3]
```diff
@@ -421,10 +423,11 @@ private XmlMapping GetMapping()
...
         public object? Deserialize(Stream stream)
         {
-            XmlTextReader xmlReader = new XmlTextReader(stream);
-            xmlReader.WhitespaceHandling = WhitespaceHandling.Significant;
-            xmlReader.Normalization = true;
-            xmlReader.XmlResolver = null;
+            //XmlTextReader xmlReader = new XmlTextReader(stream);
+            //xmlReader.WhitespaceHandling = WhitespaceHandling.Significant;
+            //xmlReader.Normalization = true;
+            //xmlReader.XmlResolver = null;
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:429]:
```suggestion
```

**Review by stephentoub — APPROVED** [review]

**Review by HongGit — APPROVED** [review]

**Review by HongGit — APPROVED** [review]

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 4]
```diff
@@ -317,9 +317,7 @@ public void Serialize(TextWriter textWriter, object? o)
         [RequiresUnreferencedCode(TrimSerializationWarning)]
         public void Serialize(TextWriter textWriter, object? o, XmlSerializerNamespaces? namespaces)
         {
-            XmlTextWriter xmlWriter = new XmlTextWriter(textWriter);
-            xmlWriter.Formatting = Formatting.Indented;
-            xmlWriter.Indentation = 2;
+            XmlWriter xmlWriter = XmlWriter.Create(textWriter, new XmlWriterSettings() { CheckCharacters = true, IndentChars = "  ", Indent = true });
             Serialize(xmlWriter, o, namespaces);
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:321]: Why explicitly set CheckCharacters and IndentChars? You are seeting them to their default values. `Indent` is the only property which you are changing from the default.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 5]
```diff
@@ -332,9 +330,7 @@ public void Serialize(Stream stream, object? o)
         [RequiresUnreferencedCode(TrimSerializationWarning)]
         public void Serialize(Stream stream, object? o, XmlSerializerNamespaces? namespaces)
         {
-            XmlTextWriter xmlWriter = new XmlTextWriter(stream, null);
-            xmlWriter.Formatting = Formatting.Indented;
-            xmlWriter.Indentation = 2;
+            XmlWriter xmlWriter = XmlWriter.Create(stream, new XmlWriterSettings() { CheckCharacters = true, IndentChars = "  ", Indent = true });
             Serialize(xmlWriter, o, namespaces);
```
Comment (mconnew) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:334]: Same for this, skip setting values to the defaults.

**Review by mconnew — CHANGES_REQUESTED** [review]
Remove the extra XmlWriterSettings property setting where you are setting the default

**Review by HongGit — APPROVED** [review]

**Review by mconnew — APPROVED** [review]

**Pull request merged**